pgms/vdist3d/vdist3d.man
.ds Vi Vista
.ds Vn 2.1
.TH vdist3d 1Vi "17 August 1994" "\*(Vi Version \*(Vn"
.SH NAME
vdist3d \- perform a 3D distance transform
.SH SYNOPSIS
\fBvdist3d \fB [-\fIoption\fP ...] [\fIinfile\fP] [\fIoutfile\fP]

.SH DESCRIPTION
\fBvdist3d\fP performs a 3D
distance transform on the \fIsource\fP image
to produce a file of a distance transformed \fIdestination\fP image.  
Each pixel in the output image contains the distance to the nearest
feature element measured in a Euclidean or Chamfer metric.
If needed, three images containing the addresses of the nearest
foreground pixels will also be produced.

.LP
The input image must be of type VBit. The output image
will be of type VFloat. If present, 
the vector images will be of type VShort.

.LP
Input and output files can be specified on the command line or
allowed to default to stdin and stdout.

.LP
If the metric is \fIeuclidean\fP, and no vectors are needed,
the algorithm used is Saito/Toriwaki's algorithm.
If the metric is \fIeuclidean\fP, but vectors are needed,
then Daniellson's algorithm is used.
For the chamfer metric the algorithm by Borgefors
is used.
Note that Daniellson's algorithm is the most expensive
(both w.r.t computing time as well as w.r.t space).
The second most expensive algorithm is Saito/Toriwaki's.
However, it is not possible to obtain the distance vectors from
this algorithm.
The cheapest is Borgefors's, which also does not yield
the distance vectors.

Note that in some rare cases, Daniellson's algorithm fails.
In that case, use Saito's algorithm (which always works)
and do without the distance vectors.


.SH LITERATURE
Toyofumi Saito, Jun-Ichiro Toriwaki,
New algorithms for euclidean distance transformation of a n-dimensional
picture with applications,
Pattern Recognition, Vol. 27, No. 11, pp 1551-1565, 1994.

Gunilla Borgefors,
"Distance Transformations in Arbitrary Dimensions",
Computer Vision, Graphics and Image Processing,
Vol 27, pp. 321-345, 1984.

P.-E. Daniellson,
"Euclidean Distance Mapping",
Computer Graphics and IMage Processing,
Vol. 14, pp. 227-248, 1980.


.SH "COMMAND LINE OPTIONS"
\fBvdist3d\fP accepts the following options in any order:
.IP \fB-help\fP 1.5i
Causes \fBvdist3d\fP to print out a message describing its options.
.IP "\fB-metric\fP \fIeuclidean/chamfer\fP"
Determines which metric to use.
.IP "\fB-vector\fP \fIyes/no\fP"
Determines whether or not a vector image will be produced.
Default: no.

.SH "SEE ALSO"
.BR Vista (7Vi)
.SH AUTHOR
Gabriele Lohmann <lohmann@cns.mpg.de>




debian/changelog
via (2.0.1svn191-0) stable; urgency=low

  *renamed via-bin to via-utils

 -- Lipsia Group (The official Lipsia developer group) <lipsia@cbs.mpg.de>  Tue, 09 Nov 2010 12:17:00 +0400 

via (2.0.0-0) stable; urgency=low

  *This version comes with the new lipsia-2.0 package
   and comprises some minor bug fixes
  *Fixed type error of VLong (is now long)

 -- Lipsia Group (The official Lipsia developer group) <lipsia@cbs.mpg.de>  Tue, 09 Nov 2010 12:17:00 +0400 

via (1.6.0-2) unstable; urgency=low

  * New internal release

 -- Frank Burkhardt <burk@cbs.mpg.de>  Mon, 20 Jul 2009 11:45:41 +0200

via (1.6.0-1) unstable; urgency=low

  * Initial release (Closes: #521328).
  * New Upstream Version. Renamed 'vcat' to 'vcatobj' to not conflict with
    atfs package in Debian.
  * Update Standards-version to 3.8.2.0 -- no changes necessary.
  * Update watchfile for recent change in upstream naming scheme.
  * Add versioned dependency on debhelper (>= 5) upon recommendation by
    lintian.

 -- Michael Hanke <michael.hanke@gmail.com>  Fri, 19 Jun 2009 13:50:52 +0200

via (1.5.2+svn3529-2) UNRELEASED; urgency=low

  * Add alternative build-dependency on atlas3-base-dev to be able to build
    the package on Debian etch.
  * FTP access to the upstream sources needs passive mode. Fixed debian/watch.

 -- Michael Hanke <michael.hanke@gmail.com>  Tue, 14 Apr 2009 02:46:10 +0200

via (1.5.2+svn3529-1) UNRELEASED; urgency=low

  * New Upstream SVN snapshot that relocates the VIA lib headers to their new
    cannonical location.

 -- Michael Hanke <michael.hanke@gmail.com>  Tue, 31 Mar 2009 22:08:01 +0200

via (1.5.2-1) UNRELEASED; urgency=low

  * Initial Debian packaging. This is a complete rewrite of the original
    Debian packaging effort done by the upstream developers.

 -- Michael Hanke <michael.hanke@gmail.com>  Thu, 26 Mar 2009 19:46:09 +0100


pgms/vdist3d/vdist3d.c
/****************************************************************
 *
 * Copyright (C) Max Planck Institute 
 * for Human Cognitive and Brain Sciences, Leipzig
 *
 * Author Gabriele Lohmann, 2004, <lipsia@cbs.mpg.de>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 *
 * $Id: vdist3d.c 3177 2008-04-01 14:47:24Z karstenm $
 *
 *****************************************************************/

/*! \brief vdist3d -- 3D distance transform

\par Description
3D distance transform.
For each background voxel, the length of the shortest
3D path to the nearest foreground voxel is computed.
The chamfer distance metric is an approximation to the
Euclidian distance.

\par References
Toyofumi Saito, Jun-Ichiro Toriwaki,
New algorithms for euclidean distance transformation of a n-dimensional
picture with applications,
Pattern Recognition, Vol. 27, No. 11, pp 1551-1565, 1994
<br>
G. Borgefors (1984).
"Distance Transforms in arbitrary dimensions",
CVGIP 27, pp.321-345.

\par Usage

        <code>vdist3d</code>

        \param -in      input image
        \param -out     output image
        \param -repn    Output pixel repn (short | float). Default: float
        \param -metric  type of metric (chamfer | euclidean). Default: euclidean

\par Examples
<br>
\par Known bugs
none.

\file vdist3d.c
\author G.Lohmann, MPI-CBS, 2004
*/


/* From the Vista library: */
#include <viaio/Vlib.h>
#include <viaio/mu.h>
#include <via.h>
#include <viaio/option.h>

/* From the standard C libaray: */
#include <stdio.h>
#include <stdlib.h>



VDictEntry MetricDict[] = {
  { "euclidean", 0 },
  { "chamfer", 1 },
  { NULL }
};

VDictEntry RepnDict[] = {
  { "short", 0 },
  { "float", 1 },
  { NULL }
};


int main (int argc, char **argv)
{
  static VLong metric = 0;
  static VLong irepn  = 1;
  static VOptionDescRec options[] = {
    { "repn", VLongRepn, 1, &irepn, VOptionalOpt, RepnDict,
      "Output pixel repn (short,float)" },
    { "metric", VLongRepn, 1, &metric, VOptionalOpt, MetricDict,
      "type of metric: chamfer, euclidean" }
  };

  FILE *in_file, *out_file;
  VAttrList list;
  VImage src, dest=NULL;
  VAttrListPosn posn;
  VRepnKind repn;
  char prg[50];	
  sprintf(prg,"vdist3d V%s", getVersion());
  fprintf (stderr, "%s\n", prg);

  /* Parse command line arguments: */
  VParseFilterCmd (VNumber (options), options,argc,argv,&in_file,&out_file);

  if (irepn == 0) repn = VShortRepn;
  else if (irepn == 1) repn = VFloatRepn;
  else VError(" <repn> must be either 'short' or 'float' ");

  
  /* Read source image(s): */
  if (! (list = VReadFile (in_file, NULL))) exit (1);

  /* Process */
  for (VFirstAttr (list, & posn); VAttrExists (& posn); VNextAttr (& posn)) {
    if (VGetAttrRepn (& posn) != VImageRepn) continue;
    VGetAttrValue (& posn, NULL, VImageRepn, & src);

    if (metric == 0) 
      dest = VEuclideanDist3d(src,NULL,repn);
    else
      dest = VChamferDist3d(src,NULL,repn);

    if (dest == NULL) exit (1);
    VSetAttrValue (& posn, NULL, VImageRepn, dest);
    VDestroyImage(src);
  }


  /* Write the results to the output file: */
  VHistory(VNumber(options),options,prg,&list,&list);
  if (VWriteFile (out_file, list))
    fprintf (stderr, "%s: done.\n",argv[0]);
  return 0;
}

